ssl/statem/extensions.c

```diff
@@ -426,11 +532,143 @@ static const EXTENSION_DEFINITION ext_defs[] = {
         TLSEXT_TYPE_psk,
         SSL_EXT_CLIENT_HELLO | SSL_EXT_TLS1_3_SERVER_HELLO
         | SSL_EXT_TLS_IMPLEMENTATION_ONLY | SSL_EXT_TLS1_3_ONLY,
+        OSSL_ECH_HANDLING_CALL_BOTH,
         NULL, tls_parse_ctos_psk, tls_parse_stoc_psk, tls_construct_stoc_psk,
         tls_construct_ctos_psk, final_psk
     }
 };
 
+#ifndef OPENSSL_NO_ECH
+/*
+ * Copy an inner extension value to outer.
+ * inner CH must have been pre-decoded into s->clienthello->pre_proc_exts
+ * already.
+ */
+static int ech_copy_inner2outer(SSL_CONNECTION *s, uint16_t ext_type,
+                                WPACKET *pkt)
+{
+    size_t ind = 0, nraws = 0;
+    RAW_EXTENSION *myext = NULL, *raws = NULL;
+
+    if (s == NULL || s->clienthello == NULL)
+        return OSSL_ECH_SAME_EXT_ERR;
+    raws = s->clienthello->pre_proc_exts;
+    if (raws == NULL)
+        return OSSL_ECH_SAME_EXT_ERR;
+    nraws = s->clienthello->pre_proc_exts_len;
+    OSSL_TRACE_BEGIN(TLS) {
+        BIO_printf(trc_out, "inner2outer: Copying ext type %d to outer\n",
+                   ext_type);
+    } OSSL_TRACE_END(TLS);
+    for (ind = 0; ind != nraws; ind++) {
+        if (raws[ind].type == ext_type) {
+            myext = &raws[ind];
+            break;
+        }
+    }
+    /*
+     * This one wasn't in inner, so re-do processing. We don't
+     * actually do this currently, but could.
+     */
+    if (myext == NULL)
+        return OSSL_ECH_SAME_EXT_CONTINUE;
+    /* copy inner value to outer */
+    if (PACKET_data(&myext->data) != NULL
+        && PACKET_remaining(&myext->data) > 0) {
+        if (!WPACKET_put_bytes_u16(pkt, ext_type)
+            || !WPACKET_sub_memcpy_u16(pkt, PACKET_data(&myext->data),
+                                       PACKET_remaining(&myext->data)))
+            return OSSL_ECH_SAME_EXT_ERR;
+    } else {
+        /* empty extension */
+        if (!WPACKET_put_bytes_u16(pkt, ext_type)
+            || !WPACKET_put_bytes_u16(pkt, 0))
+            return OSSL_ECH_SAME_EXT_ERR;
+    }
+    return 1;
+}
+
+/*
+ * DUPEMALL is useful for testing - this turns off compression and
+ * causes two calls to each extension constructor, which'd be the same
+ * as making all entries in ext_tab use the CALL_BOTH value
+ */
+# undef DUPEMALL
+
+/*
+ * Check if we're using the same/different key shares
+ * return 1 if same key share in inner and outer, 0 otherwise
+ */
+int ossl_ech_same_key_share(void)
```

Comment:
Currently, yes, but it's a compile time setting from the `ext_defs.ech_handling` field so we should keep this. (If the same key share is in both inner and outer then we need to up-ref the EVP pointer.)

Having the ability to use different inner/outer key shares is also good for testing to ensure our server code can handle that in case some client does that. (Not too likely today.)  I could also just about imagine some argument for a security/interop benefit in using different key shares could arise in future.